### PR TITLE
 Fixes codership/mysql-wsrep#141 and codership/mysql-wsrep#143 - 5.6 branch 

### PR DESCRIPTION
**Pull request by ayurchen**
codership/mysql-wsrep#141 and codership/mysql-wsrep#143 - 5.6 branch

### CODE REVIEW

**Comment by philip-galera** [comment]
Please give me an hour or two to test this ...

**Comment by temeo** [comment]
MTR run started failing at some point with

```
2015-06-08 10:23:06 28049 [Note] WSREP: Running: 'wsrep_sst_rsync --role 'joiner' --address '127.0.0.2:15007' --datadir '/var/lib/jenkins/workspace/mtr-galera-mysql-5.6/mysql-wsrep/mysql-test/var/1/mysqld.2/data/' --defaults-file '/var/lib/jenkins/workspace/mtr-galera-mysql-5.6/mysql-wsrep/mysql-test/var/1/my.cnf' --defaults-group-suffix '.2' --parent '28049'  '' '
WSREP_SST: [ERROR] rsync daemon port '15007' has been taken (20150608 10:23:06.880)
WSREP_SST: [INFO] Joiner cleanup. rsync PID: 0 (20150608 10:23:06.885)
WSREP_SST: [INFO] Joiner cleanup done. (20150608 10:23:06.892)
2015-06-08 10:23:06 28049 [ERROR] WSREP: Failed to read 'ready <addr>' from: wsrep_sst_rsync --role 'joiner' --address '127.0.0.2:15007' --datadir '/var/lib/jenkins/workspace/mtr-galera-mysql-5.6/mysql-wsrep/mysql-test/var/1/mysqld.2/data/' --defaults-file '/var/lib/jenkins/workspace/mtr-galera-mysql-5.6/mysql-wsrep/mysql-test/var/1/my.cnf' --defaults-group-suffix '.2' --parent '28049'  '' 
    Read: '(null)'
2015-06-08 10:23:06 28049 [ERROR] WSREP: Process completed with error: wsrep_sst_rsync --role 'joiner' --address '127.0.0.2:15007' --datadir '/var/lib/jenkins/workspace/mtr-galera-mysql-5.6/mysql-wsrep/mysql-test/var/1/mysqld.2/data/' --defaults-file '/var/lib/jenkins/workspace/mtr-galera-mysql-5.6/mysql-wsrep/mysql-test/var/1/my.cnf' --defaults-group-suffix '.2' --parent '28049'  '' : 16 (Device or resource busy)
2015-06-08 10:23:06 28049 [ERROR] WSREP: Failed to prepare for 'rsync' SST. Unrecoverable.
```

**Comment by philip-galera** [comment]
With this

```
perl mysql-test-run.pl --suite galera galera_sst_xtrabackup-v2-options --big-test
```

I get:

```
/home/philips/git/mysql-wsrep-5.6-gh141-5.6/scripts/wsrep_sst_common: line 148: AUTH_VEC[1]: unbound variable
```

**Comment by philip-galera** [comment]
Here is what is passed to the script:

```
WSREP_SST_OPT_AUTH=root:
```

So the password is empty, which confuses the script. Should I fix the test or the SST should be fixed?

**Comment by temeo** [comment]
Also it looks like stale rsync daemon is left behind if joiner side sst fails for some reason.

```
WSREP_SST: [ERROR] Parent mysqld process (PID:6292) terminated unexpectedly. (20150608 10:43:11.568)
WSREP_SST: [INFO] Joiner cleanup. rsync PID: 0 (20150608 10:43:11.570)
WSREP_SST: [INFO] Joiner cleanup done. (20150608 10:43:11.573)
```

Script tries to kill PID 0.

**Comment by temeo** [comment]
Running upgrade test with gh141-5.6 branch fails quickly with hanging joiner. This is because donor side error:

```
2015-06-08 11:24:06 492 [Note] WSREP: Shifting SYNCED -> DONOR/DESYNCED (TO: 1603)
2015-06-08 11:24:06 492 [Note] WSREP: IST request: be96ee5b-0dd0-11e5-b709-065383db6e64:18-1580|tcp://172.17.9.75:10012
2015-06-08 11:24:06 492 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2015-06-08 11:24:06 492 [ERROR] WSREP: sst_donate_other(): appending auth env failed: -2
2015-06-08 11:24:06 492 [Note] WSREP: async IST sender starting to serve tcp://172.17.9.75:10012 sending 19-1580
```

There are two bugs here:
- Error appending auth env (no clear indication why it happens)
- Error is not handled properly by galera side

File: sql/wsrep_sst.cc [hunk 1]
```diff
@@ -498,8 +529,21 @@ static ssize_t sst_prepare_other (const char*  method,
...
+
+  if ((ret= sst_append_auth_env(env, sst_auth)))
+  {
+    WSREP_ERROR("sst_prepare_other(): appending auth failed: %d", ret);
+    return ret;
+  }
+
   pthread_t tmp;
-  sst_thread_arg arg(cmd_str);
+  sst_thread_arg arg(cmd_str(), env());
```
Comment (temeo) [inline sql/wsrep_sst.cc:546]: Not quite part of this patch, but below return value from `pthread_create()` is returned directly. According to http://man7.org/linux/man-pages/man3/pthread_create.3.html the error returned is error number (positive integer). However, all other error handling in this function return negative error values.

File: sql/wsrep_utils.cc [hunk 2]
```diff
@@ -89,6 +89,91 @@ wsrep_prepend_PATH (const char* path)
...
+
+        if (env_[len_])
+        {
+            ++len_;
+            env_[len_] = NULL;
+        }
+    }
+
+    /* if either realloc() or strdup() failed, errno had been set */
+    errno_ = errno;
```
Comment (temeo) [inline sql/wsrep_utils.cc]: Here it would be better to check if error actually happened before assigning errno_.

File: sql/wsrep_sst.cc [hunk 3]
```diff
@@ -1085,15 +1119,29 @@ wsrep_cb_status_t wsrep_sst_donate_cb (void* app_ctx, void* recv_ctx,
...
+  {
+    WSREP_ERROR("sst_donate_other(): env var ctor failed: %d", -env.error());
+    return WSREP_CB_FAILURE;
+  }
+
   int ret;
+  if ((ret= sst_append_auth_env(env, sst_auth_real)))
+  {
+    WSREP_ERROR("sst_donate_other(): appending auth env failed: %d", ret);
+    return WSREP_CB_FAILURE;
```
Comment (temeo) [inline sql/wsrep_sst.cc:1133]: Error logging above: This function is not `sst_donate_other()`.

**Comment by temeo** [comment]
This pull request still fails tests `galera.galera_ist_xtrabackup-v2`, `galera.galera_ist_innodb_flush_logs`. Both have the same error in `innobackup.backup.log`:

```
Option user requires an argument
innobackupex: got a fatal error with the following stacktrace: at /usr/bin/innobackupex line 3860.
        main::check_args() called at /usr/bin/innobackupex line 1522
innobackupex: Error: Bad command line arguments
```

**Comment by philip-galera** [comment]
The WSREP_SST_OPT_AUTH variable is empty .

**Comment by ayurchen** [comment]
Is wsrep_sst_auth set in any of [mysqld] or [sst] sections of joiner's my.cnf?

**Comment by philip-galera** [comment]
No, wsrep_sst_auth is not set anywhere and yet all tests except galera.galera_ist_xtrabackup-v2, galera.galera_ist_innodb_flush_logs work.

**Comment by philip-galera** [comment]
Should I fix this by setting it in the test? Or there is more to be done -- I wonder why it worked previously...

**Comment by ayurchen** [comment]
I think it didn't really... rsync does not need / ignores wsrep_sst_auth, but both mysqldump and xtrabackup do (and, strictly speaking, both need it on the donor side, however it makes little sense to let the node join the cluster without it set)

**Comment by philip-galera** [comment]
Ok, will fix the tests and push into that tree. Though for the record one must note that the error message was not helpful when debugging the problem.

**Comment by temeo** [comment]
Why is not WSREP_SST_OPT_AUTH set if the user is passed to SST script through environment? Why it works for mysqldump SST?

**Comment by philip-galera** [comment]
The mysqldump tests explicitly set wsrep_sst_auth . The other tests do not and rsync is the default mechanism for almost all tests.

**Comment by ayurchen** [comment]
WSREP_SST_OPT_ATUH is passed through environment. But if wsrep_sst_auth is not set in [mysqld] it will be empty. And then the script will try to read the same from [sst]. If it is not set there, then both user and password will be empty. ~~Perhaps mysqldump could connect to server with default user/password or read it from [mysql] section, and maybe from another my.cnf.~~
Also note that xtrabackup-v2 script was updated in this pull request, so it may behave a tad differently (and hopefully more correctly) now.

**Comment by philip-galera** [comment]
Tests fixed. Should I merge this pull request?

**Comment by temeo** [comment]
If it passes all MTR tests it should be good enough to merge.

**Pull request merged**